### PR TITLE
Reduce indirection when executing sync integrations

### PR DESCRIPTION
**Pull request by edufschmidt**
This PR attempts to simplify and reduce indirection within the sync module.

The concepts of instance, pipeline, and segment have been removed, and their functionality moved either to the Sync class or to an `utils.ts` script.

The `httpRequest` method used only in the `oauth.ts` script was removed, as it was redundant with another implementation in the same repo (which was also updated to make use of `axios-retry`).

All tests previously defined in `instance.spec.ts` have been adapted and are passing.

The next step is to simplify the execution context to which integration code has access, ideally by providing it with a Kernel instance and read-only credentials, as discussed [here](https://jel.ly.fish/improvement-simplify-context-available-sync-integrations-9fe69aa).

Change-type: patch

### CODE REVIEW

**Comment by edufschmidt** [comment]
@LucianBuzzo this is still a WIP, but I thought you would be interested in taking a look and sharing some thoughts.

File: lib/index.ts [hunk 1]
```diff
@@ -318,13 +320,7 @@ export class Worker {
...
 				);
 			},
-			enqueueAction: (...args) => {
-				return this.enqueueAction(...args);
-			},
-			cards: {
-				...CORE_CARDS,
-				...self.getTypeContracts(),
-			},
+			enqueueAction: this.enqueueAction,
```
Comment (LucianBuzzo) [inline lib/index.ts:323]: Does `this` need to be bound for the assignment to work correctly? 🤔

**Review by LucianBuzzo — COMMENTED** [review]
This all looks a lot cleaner 👍  nice work!

File: lib/sync/context.ts [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+		const sessionCard = await workerContext.getCardById(session, session);
+		if (!sessionCard) {
+			return null;
+		}
+		// TODO: Replace this return type with the session contract interface
+		return sessionCard.data.actor as string;
+	};
+
+	const contextObject: SyncActionContext = {
+		logger: {
```
Comment (Ereski) [inline lib/sync/context.ts:129]: IMO this makes more sense either with these functions as part of the root level of the object, or with `logger` renamed to `log` so:

```
context.log.error(...)
```

instead of

```
context.logger.error(...)
```

File: lib/sync/index.ts [hunk 3]
```diff
@@ -357,53 +356,291 @@ export class Sync {
...
+	 * @param {Object} contract - action target contract
 	 * @param {Object} context - execution context
 	 * @param {Object} options - options
 	 * @param {String} options.actor - actor id
 	 * @param {String} options.timestamp - timestamp
 	 * @param {String} [options.origin] - OAuth origin URL
 	 * @returns {Object[]} inserted cards
 	 */
-	async translate(
+	translate = async (
```
Comment (Ereski) [inline lib/sync/index.ts:622]: why this change?

File: lib/sync/types.ts [hunk 4]
```diff
@@ -2,28 +2,79 @@ import type { LogContext } from '@balena/jellyfish-logger';
 import type { Contract } from '@balena/jellyfish-types/build/core';
 import type { Operation } from 'fast-json-patch';
 import type { Map } from '../types';
-import type { SyncActionContext } from './sync-context';
+import type { Logger, SyncActionContext } from './context';
+export interface IntegrationDefinition {
```
Comment (Ereski) [inline lib/sync/types.ts:6]: whitespace separating the import list and the body of the file is good

File: lib/sync/utils.ts [hunk 5]
```diff
@@ -0,0 +1,329 @@
+import axios, { AxiosError, Method } from 'axios';
+import jsone from 'json-e';
+import * as assert from '@balena/jellyfish-assert';
+import * as errors from './errors';
+import _ from 'lodash';
+import { SyncActionContext } from './context';
+import { Contract } from '@balena/jellyfish-types/build/core';
+import axiosRetry from 'axios-retry';
+export interface HttpRequestOptions {
```
Comment (Ereski) [inline lib/sync/utils.ts:9]: same here

File: lib/sync/context.ts (hunk 2 above)
Comment (edufschmidt) [reply lib/sync/context.ts:129]: Hmm, won't that make naming inconsistent? AFAIK logging methods are derived from a `Logger` interface obtained through a `getLogger` method. What's the motivation for naming it `log` here or for making these functions part of the root level object?

File: lib/index.ts (hunk 1 above)
Comment (edufschmidt) [reply lib/index.ts]: This change was reverted @LucianBuzzo.
